esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -1241,194 +1241,205 @@ void setDeleteFileResults(const char* fileName, const char* nodeGroup, bool fail
     actionResults.append(*resultObj.getClear());
 }
 
-typedef enum {
-    DeleteActionSuccess,
-    DeleteActionFailure,
-    DeleteActionSkip
-} DeleteActionResult;
-
+void cleanLogicalFileNames(StringArray &fileNames, StringArray &uniqueFileNames)
+{
+    ForEachItemIn(i, fileNames)
+    {
+        const char *fn = fileNames.item(i);
+        if (isEmptyString(fn))
+            continue;
 
-DeleteActionResult doDeleteFile(const char *fn, IUserDescriptor *userdesc, StringArray &superFiles, StringArray &failedFiles,
-    const char *auditStr, StringBuffer& returnStr, IArrayOf<IEspDFUActionInfo>& actionResults, bool superFilesOnly, bool removeFromSuperfiles, bool deleteRecursively);
+        uniqueFileNames.appendUniq(fn);
+    }
+}
 
-bool doRemoveFileFromSuperfiles(const char *lfn, IUserDescriptor *userdesc, StringArray &superFiles, StringArray &failedFiles, bool deleteRecursively,
-    const char *auditStr, StringBuffer& returnStr, IArrayOf<IEspDFUActionInfo>& actionResults)
+bool CWsDfuEx::DFUDeleteFiles(IEspContext &context, IEspDFUArrayActionRequest &req, IEspDFUArrayActionResponse &resp)
 {
-    StringArray emptySuperFiles;
-    IDistributedFileDirectory &fdir = queryDistributedFileDirectory();
+    class CDFUDeleteFilesHelper
     {
-        Owned<IDistributedFile> df = fdir.lookup(lfn, userdesc, true, false, false, nullptr, defaultPrivilegedUser);
-        if(!df)
-            return false;
-        Owned<IDistributedSuperFileIterator> supers = df->getOwningSuperFiles();
-        ForEach(*supers)
+        IEspContext &context;
+        Owned<IUserDescriptor> userdesc;
+        IArrayOf<IEspDFUActionInfo> actionResults;
+        StringBuffer returnStr, auditStr;
+        StringAttr espProcess;
+
+        void cleanOwningSuperFiles(IDistributedFileDirectory &fdir, StringArray &fileNames, bool removeFromSuperFiles, bool removeEmptyOwningSuperFiles)
+        {
+            StringArray emptyOwningSuperFiles;
+            ForEachItemInRev(i, fileNames)
+            {
+                CDfsLogicalFileName dfsLFN;
+                dfsLFN.set(fileNames.item(i));
+
+                const char *lfn = dfsLFN.get();
+                Owned<IDistributedFile> df = fdir.lookup(lfn, userdesc, true, false, false, nullptr, defaultPrivilegedUser);
+                if (!df)
+                {
+                    VStringBuffer message("Could not delete file %s: file not found.", lfn);
+                    setResults(lfn, nullptr, true, message);
+                    fileNames.remove(i);
+                    continue;
+                }
+                Owned<IDistributedSuperFileIterator> supers = df->getOwningSuperFiles();
+                if (!supers->first())
+                    continue;
+
+                if (!removeFromSuperFiles)
```

Comment:
> just because it is a super in the delete list, doesn't mean it's not owned by others too,

Do you mean that the super may be owned by another super? Then, should I change this block of code to:
if (!removeFromSuperFiles)
{
    VStringBuffer message("The file %s is not deleted because it is owned by superfile.", lfn);
    setResults(lfn, nullptr, true, message);
    fileNames.remove(i);
    continue;
}

Reply:
I'd change it to:
```
    Owned<IDistributedSuperFileIterator> superOwners = df->getOwningSuperFiles();
    if (!removeFromSuperFiles && superOwners->first())
    {
        VStringBuffer message("The file %s can not be deleted because it is owned by one or more superfiles.", lfn);
        setResults(lfn, nullptr, true, message);
        continue;
    }
```

see other comment re. single iteration of logicalFilenames, in which case you don'tneed to remove items from list.

Reply:
Yes, you are correct. I am going to move most of the code lines from the cleanOwningSuperFiles() to the deleteFiles().